### ISSUE TITLE
Allow injection of JPA properties that contain object instances

### ISSUE THREAD

**Issue by jeroenvs** [spring-projects/spring-boot#2808]
We want our entity manager factory to use the LocalValidatorFactoryBean, allowing our constraint validators to access the application context. However, this does not work automatically with spring boot (in my project).

It is possible to inject a custom validator into the entity manager factory by providing a JPA property:

```
    properties.put("javax.persistence.validation.factory", validator());
```

And I figured out that you can provide this same property in Spring boot with the application.properties. But this only supports string values, while I need a bean reference. Is it possible to do bean references in the application.properties?

As work-around I currently create a custom entity manager factory in my application config:

```
@Bean
public LocalContainerEntityManagerFactoryBean entityManagerFactory(EntityManagerFactoryBuilder builder) {
    Map<String, Object> properties = new HashMap<>();
    properties.put("hibernate.ejb.naming_strategy", ConventionNamingStrategy.class.getName());
    properties.put("hibernate.dialect", ImprovedHsqlDialect.class.getName());
    properties.put("hibernate.hbm2ddl.auto", "validate");
    properties.put("hibernate.jdbc.use_get_generated_keys", true);
    properties.put("hibernate.id.new_generator_mappings", true);
    properties.put("hibernate.generate_statistics", false);
    properties.put("javax.persistence.validation.factory", validator());

    return builder.dataSource(dataSource)
                  .packages(ApplicationConfig.class)
                  .properties(properties)
                      .build();
}
```

But this is not entirely desirable as I want to configure this in my application.properties

Is there a better way to inject a custom validation factory in my entity manager factory? Preferably while keeping the automatic entity manager factory?

**Comment by philwebb** [comment]
I don't think that there is currently an easy way to do this. We'll take a look and see if we can add something for 1.3

**Comment by philwebb** [comment]
Pushing back to 1.4 since we're running out of time and there is a work-around.

**Comment by stevestorey** [comment]
There's now an easier workaround, where you can define a configuration class like:
```
@Configuration
public class HibernateValidationConfiguration extends HibernateJpaAutoConfiguration {

    @Autowired private Validator validator;

    @Override
    protected void customizeVendorProperties(Map<String, Object> vendorProperties) {
        super.customizeVendorProperties(vendorProperties);
        vendorProperties.put("javax.persistence.validation.factory", validator);
    }
}
```

and then just ```@Import(HibernateValidationConfiguration.class)``` into a configuration where the validator is defined etc (or you can have the main class extend HibernateJpaAutoConfiguration directly)

**Comment by jeroenvs** [comment]
Thanks! That looks like a neat solution

**Comment by vyletpet** [comment]
The mentioned workaround is not working as of SpringBoot 2.0.
Can you please take a look?
Anyway what is now the proper way of customizing Hibernate properties?

**Comment by snicoll** [comment]
@vyletpet asking questions on a closed issue is not a great way to get support. Ask on StackOverflow or join us on Gitter. Also, [reading the release notes](https://github.com/spring-projects/spring-boot/wiki/Spring-Boot-2.0.0-RC1-Release-Notes#hibernate-properties-customization) should help.

**Comment by snicoll** [comment]
Extending our auto-configuration is a bad idea. We've actually implemented that in #11211

Duplicate of #11211

**Comment by shenyu1997** [comment]
You can do like this if you are using Spring Boot 2.1.0+:


    @Configuration
    @Lazy
    class SpringValidatorConfiguration {

        @Autowired
        private Validator validator;

        @Bean
        @Lazy
        public HibernatePropertiesCustomizer hibernatePropertiesCustomizer(final Validator validator) {
            return new HibernatePropertiesCustomizer() {

                @Override
                public void customize(Map<String, Object> hibernateProperties) {
                    hibernateProperties.put("javax.persistence.validation.factory", validator);
                }
            };
        }
    }

The idea from  [Spring Boot 2.0.0 M6 - Add Hibernate Interceptor][1]


and [Spring Boot - Hibernate custom constraint doesn't inject Service][2]


  [1]: https://stackoverflow.com/questions/47465728/spring-boot-2-0-0-m6-add-hibernate-interceptor/48207856#48207856
  [2]: https://stackoverflow.com/questions/50212117/spring-boot-hibernate-custom-constraint-doesnt-inject-service/50213178#50213178
